### PR TITLE
Add functions to retrieve field index holes

### PR DESCRIPTION
**Pull request by lbschanno**
Add functionality to retrieve field index holes for index rows and reversed index rows from the metadata table.

Required to support work for #825.

### CODE REVIEW

File: src/main/java/datawave/query/util/AllFieldMetadataHelper.java [hunk 1]
```diff
@@ -1050,6 +1058,348 @@ public Set<String> loadDatatypes() throws TableNotFoundException {
...
+                prevFieldName = fieldName;
+                prevDatatype = datatype;
+                prevDate = date;
+                startDate = date;
+                prevColumnFamily = columnFamily;
+                continue;
+            }
+            
+            // If the column family is different, determine record the last date range, and begin collecting date ranges for the next batch of related rows.
+            if (!prevColumnFamily.equals(columnFamily)) {
```
Comment (ivakegg) [inline src/main/java/datawave/query/util/AllFieldMetadataHelper.java:1139]: I think you may need to verify that you are in the same fieldname.  I could see several field names in a row only with F fields and not being indexed.  I think this whole loop might be cleaner if you simply gathered all the data for a fields, then when you get to the next field (or the end) you update the field index wholes.

File: src/test/java/datawave/query/util/AllFieldMetadataHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+            }
+        };
+        
+        private Supplier<Map<String,Map<String,FieldIndexHole>>> getIndexHoleFunction(String cf) {
+            return cf.equals("i") ? INDEX_FUNCTION : REVERSED_INDEX_FUNCTION;
+        }
+        
+        /**
+         * Test against data that has no field index holes.
+         */
```
Comment (ivakegg) [inline src/test/java/datawave/query/util/AllFieldMetadataHelperTest.java:130]: Need a test where we have many fields in a row that only have 'f' entries and no 'i' entries.

Comment (lbschanno) [reply src/test/java/datawave/query/util/AllFieldMetadataHelperTest.java:130]: A test like that is already present in `testAllDatesAreIndexHoles()` [here](https://github.com/NationalSecurityAgency/datawave-metadata-utils/pull/29/files/f2bc5377d7548f9f18283852ff42727a398efd53#diff-da568db7f6796bd73bbb2e23b7124f99a8feb9ce24dbfff670a54fbeb5bc1e13R306), where multiple 'f' entries are created for different fields in a row with no 'i' entries.

File: src/main/java/datawave/query/util/AllFieldMetadataHelper.java (hunk 1 above)
Comment (lbschanno) [reply src/main/java/datawave/query/util/AllFieldMetadataHelper.java:1139]: This check is performed [here](https://github.com/NationalSecurityAgency/datawave-metadata-utils/pull/29/files/f2bc5377d7548f9f18283852ff42727a398efd53#diff-940dceae6ce6aaacc9b6c6f5ff78b5508669273ae516eff29ab03ec0efa93aebR1164) if we know that the column family has not changed. I will try to make the documentation surrounding that check a little clearer.

Comment (lbschanno) [reply src/main/java/datawave/query/util/AllFieldMetadataHelper.java:1143]: I have updated the documentation to improve the clarity.

**Review by ivakegg — APPROVED** [review]

**Pull request merged**